pyadjoint/checkpointing.py

```diff
@@ -180,12 +185,19 @@ def _(self, cp_action, timestep):
                 for deps in self.tape.timesteps[timestep - 1].checkpointable_state:
                     self._global_deps.add(deps)
             else:
-                deps_to_clear = self._global_deps - self._global_deps.intersection(
-                    self.tape.timesteps[timestep - 1].checkpointable_state)
+                # Check if the block variables stored at `self._global_deps` are still
+                # dependencies in the previous timestep. If not, will remove them from the
+                # global dependencies.
+                deps_to_clear = self._global_deps.difference(self._global_deps.intersection(
```

Comment:
I don't think you need to have the `intersection` here.

Reply:
But I could be wrong.

Reply:
I will check this.